### ISSUE TITLE
Ruby < 2.4 fail to build on Fedora 26

### ISSUE THREAD

**Issue by SebastianPoehn** [rvm/rvm#4057]
## Description

Fedora 26 introduced openssl 1.1 whose API is not compatible with 1.0

## Steps to reproduce

1. Default Fedora 26 installation.
2. rvm install 2.2.6

## Expected behavior

Build fails because of a lot of openssl related build fails.

## How to Fix

For backward compat there is compat-openssl10-devel.

So we should:
if Fedora 26 && ruby < 2.4
   compat-openssl10-devel
else
  openssl-devel

After doing the change (unconditionally) the build works fine for me and I can use 2.2.6

**Comment by pkuczynski** [comment]
I prepared a fix for this issue. Could you please run following commands and see f it works?

```
apt-get remove compat-openssl10-devel
rvm get branch /bugfix/4057-fedora-26-compat-openssl10-devel
rvm install 2.2.6
```

Thanks for help!

**Comment by SebastianPoehn** [comment]
Thanks for that prompt reply.

Unfortunately the $1 is not set for me.

Created a PR that makes it work. I can do a final test after you have merged to your bugfix branch.

https://github.com/rvm/rvm/pull/4062

**Comment by pkuczynski** [comment]
This PR would not work because `${rvm_ruby_version}` shows only pure version without specifing ruby interpreter. It means it might conflict in the future with other Ruby interpreters like `ree`, `rbx` and others.

**Comment by pkuczynski** [comment]
Possibli there is an error somewhere else, as this PR contains quite a lot of changes. Could you please gist output of:

```
apt-get remove compat-openssl10-devel
rvm get branch /bugfix/4057-fedora-26-compat-openssl10-devel
rvm install 2.2.6 --debug
```

**Comment by SebastianPoehn** [comment]
<script src="https://gist.github.com/SebastianPoehn/4a4ab027d243f17684df864371aeebd6.js"></script>

**Comment by pkuczynski** [comment]
Could you please show me output of `rvm info`?

**Comment by ANeilan** [comment]
running fedora 26, it still tries to install openssl-devel. using branch /bugfix/4057-fedora-26-compat-openssl10-devel 

https://gist.github.com/ANeilan/a5b3ddc9091243ffb9a822bafe062cd4

rvm info
https://gist.github.com/ANeilan/93646cc5998c8fbf6da91f579bf2c859

**Comment by pkuczynski** [comment]
@ANeilan I would need the same type of log as @SebastianPoehn posted, but from the latest branch...

**Comment by ANeilan** [comment]
https://gist.github.com/ANeilan/03c1f35cfcbd4afb08522c84d27142c7

here's with rvm latest branch

**Comment by ANeilan** [comment]
and here is the actual build where it still fails

https://gist.github.com/ANeilan/846b52c46d9136e8a23b9e40d7bd32a8

**Comment by pkuczynski** [comment]
@ANeilan and @SebastianPoehn I just made another fix and it should work now. Could you please test it:

```
apt-get remove compat-openssl10-devel
rvm get branch /bugfix/4057-fedora-26-compat-openssl10-devel
rvm install 2.2.6 --debug
```

**Comment by SebastianPoehn** [comment]
Perfect. Not it works

**Comment by pkuczynski** [comment]
Perfect. I will merge it in then and you can switch over to `rvm get master`

**Comment by ANeilan** [comment]
2.2 builds, but 2.3.x still won't build

**Comment by pkuczynski** [comment]
@ANeilan could you please create a separate issue for 2.3 and fill all the information requested in issue template (especially all logs)

**Comment by AnwarShah** [comment]
The problem is still there. I just get rvm master and it seems the error is different now. It's hanging at 
```
* Error in `../../miniruby': corrupted double-linked list: 0x0000000001cbf8e0 ***
...*** Error in `../../miniruby': corrupted double-linked list: 0x0000000001a3b510 ***
```

**Comment by dioni21** [comment]
@AnwarShah Also not working for ruby-2.2.5 on fedora 26, even with rvm-master.  Is there any newer version other than "rvm get master"?

To test:
gem install bundler
bundle install

In any project I have.

**Comment by AnwarShah** [comment]
@dioni21 Hi, the fix was actually patching ruby. The patch is this https://bugs.ruby-lang.org/attachments/6655/ruby_2_3_gcc7.patch. I found it on another rbenv thread. 

To patch, you need to pass a special formatting to `-p` key. Check the rvm install help thoroughly.

**Comment by dioni21** [comment]
@AnwarShah Should'nt this patch be included into RVM default patches?  Like those in ~/.rvm/patches/ruby/...

**Comment by Reedlee** [comment]
@AnwarShah I made next:
`
rvm get master
sudo yum install -y compat-openssl10-devel --allowerasing
curl -O https://bugs.ruby-lang.org/attachments/download/6655/ruby_2_3_gcc7.patch 
mv ruby_2_3_gcc7.patch ~/.rvm/patches/ruby/2.3.4
rvm install 2.3.3 --patch ruby_2_3_gcc7.patch%0
`
After all I have installed ruby 2.3.4

**Comment by cben** [comment]
ref: patch origin is https://bugs.ruby-lang.org/issues/13150
Expected to be included in ruby 2.3.5, but sounds useful for RVM to ship patch so it can build 2.3.0 .. 2.3.4.

**Comment by Dylan-M** [comment]
This actually breaks rvm's multiple version capability, since you can't have Ruby < 2.4 and Ruby 2.4+ installed and working at the same time since they require conflicting libraries. I wouldn't consider this a true fix.

To wit, compat-openssl10-devel and openssl-devel cannot be installed at the same time as they're conflicting libraries:
```bash
# yum install -y openssl-devel
Last metadata expiration check: 0:32:57 ago on Wed 13 Dec 2017 07:32:36 AM EST.
Error:
 Problem: problem with installed package compat-openssl10-devel-1:1.0.2m-1.fc27.x86_64
  - package compat-openssl10-devel-1:1.0.2m-1.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0f-9.fc27.i686
  - package compat-openssl10-devel-1:1.0.2j-9.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0f-9.fc27.i686
  - conflicting requests
  - package compat-openssl10-devel-1:1.0.2m-1.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0g-1.fc27.i686
  - package compat-openssl10-devel-1:1.0.2j-9.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0g-1.fc27.i686
  - package compat-openssl10-devel-1:1.0.2m-1.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0f-9.fc27.x86_64
  - package compat-openssl10-devel-1:1.0.2j-9.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0f-9.fc27.x86_64
  - package compat-openssl10-devel-1:1.0.2m-1.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0g-1.fc27.x86_64
  - package compat-openssl10-devel-1:1.0.2j-9.fc27.x86_64 conflicts with openssl-devel provided by openssl-devel-1:1.1.0g-1.fc27.x86_64
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages)
```

Which, from a user perspective looks like this:
```bash
$ rvm install 2.4.0
Searching for binary rubies, this might take some time.
No binary rubies available for: fedora/27/x86_64/ruby-2.4.0.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for fedora.
Installing requirements for fedora.
Installing required packages: openssl-devel.....
Error running 'requirements_centos_libs_install openssl-devel',
please read /home/<user>/.rvm/log/1513169785_ruby-2.4.0/package_install_openssl-devel.log
Requirements installation failed with status: 1.
```

I had, for this user, initially installed Ruby 2.4.1. Then I had installed the compat lib for openssl so that they could install 2.3.1. This results in the following from rvm:
```bash
$ rvm list

rvm rubies

=* ruby-2.3.1 [ x86_64 ]
   ruby-2.4.1 [ broken ]

# => - current
# =* - current && default
#  * - default
```

Since we're doing development that has to be tested against multiple versions of Ruby, this is a pretty big problem for us.
